src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -68,119 +71,132 @@ struct Partition {
+  // For a value seen on the probe side, return an iterator pair into the matching positions on the build side
   template <typename InputType>
-  const std::vector<SmallPosList>::const_iterator find(const InputType& value) const {
-    DebugAssert(_mode == JoinHashBuildMode::AllPositions, "find is invalid for SinglePosition mode, use contains");
+  const std::pair<RowIDPosList::const_iterator, RowIDPosList::const_iterator> find(const InputType& value) const {
+    DebugAssert(_mode == JoinHashBuildMode::AllPositions, "find is invalid for ExistenceOnly mode, use contains");
+    DebugAssert(_unified_pos_list, "_unified_pos_list not set - was finalize called?");
 
     const auto casted_value = static_cast<HashedType>(value);
-    if (!_values) {
-      const auto hash_table_iter = _hash_table.find(casted_value);
-      if (hash_table_iter == _hash_table.end()) return end();
-      return _pos_lists.begin() + hash_table_iter->second;
-    } else {
-      const auto values_iter =
-          std::find_if(_values->begin(), _values->end(), [&](const auto& pair) { return pair.first == casted_value; });
-      if (values_iter == _values->end()) return end();
-      return _pos_lists.begin() + values_iter->second;
+    const auto hash_table_iter = _offset_hash_table.find(casted_value);
+
+    if (hash_table_iter == _offset_hash_table.end()) {
+      // Not found, return an empty range
+      return {_unified_pos_list->pos_list.end(), _unified_pos_list->pos_list.end()};
     }
+
+    return {_unified_pos_list->pos_list.begin() + _unified_pos_list->offsets[hash_table_iter->second],
+            _unified_pos_list->pos_list.begin() + _unified_pos_list->offsets[hash_table_iter->second + 1]};
   }
 
   // For a value seen on the probe side, return whether it has been seen on the build side
   template <typename InputType>
   bool contains(const InputType& value) const {
     const auto casted_value = static_cast<HashedType>(value);
-
-    if (!_values) {
-      return _hash_table.find(casted_value) != _hash_table.end();
-    } else {
-      const auto values_iter =
-          std::find_if(_values->begin(), _values->end(), [&](const auto& pair) { return pair.first == casted_value; });
-      return values_iter != _values->end();
-    }
+    return _offset_hash_table.find(casted_value) != _offset_hash_table.end();
   }
 
-  const std::vector<SmallPosList>::const_iterator begin() const { return _pos_lists.begin(); }
-
-  const std::vector<SmallPosList>::const_iterator end() const { return _pos_lists.end(); }
-
  private:
-  HashTable _hash_table;
-  std::vector<SmallPosList> _pos_lists;
-  JoinHashBuildMode _mode;
-  std::optional<std::vector<std::pair<HashedType, Offset>>> _values{std::nullopt};
+  // During the build phase, the small_vectors cause many small allocations. Instead of going to malloc every time,
+  // we create our own (non-thread-safe) pool, which is discarded once finalize() is called
```

Comment:
> During the build phase, the small_vectors cause many small allocations. Instead of going to malloc every time,
we create our own pool, which is discarded once finalize() is called. The pool is unsynchronized (i.e., non-thread-
safe) by design. This way, we can quickly perform a high number of allocations without having to synchronize with
other threads for each allocation. Instead, we synchronize only when we refill the underlying
monotonic_buffer_resource. This works because each PosHashTable is used by exactly one thread.

Reply:
Works for me.